feat/add-login-page

Dewhs opened the pull request:
Just add basic login ui page and basic navBar and navigation system

Dewhs committed:
feat: add custom fontFamily

Dewhs committed:
feat: update app icon (work with adaptative android color theme)

Dewhs committed:
feat: update theme

Dewhs committed:
feat: add squircle library

Dewhs committed:
feat: add login page

note: the next steps are to implement a routing system and use the appropriate architecture pattern to integrate the front with the api.

Dewhs committed:
fix: rename squircleShape -> shape + remove useless .clip() + reformat

Dewhs committed:
fix(CheckBox) : when not checked just hidde Icon istead of remove from graphic tree

Dewhs committed:
fix: BasicText() -> Text(), use Material Text() composable to auto handle textColor in dynamicDarkColorScheme

Dewhs committed:
fix: Button -> MButtonPrimary

this name can evolved to just MButton. But for now I choose to named Primary to refactor later (if we need to) when we will need others types of button

RomanJos reviewed (approved):
Ben nickel tout ça

Dewhs committed:
feat: set basic navigation

Dewhs committed:
feat: add navBar (except for login) with 4 Screens

(using Scaffold it fix darkMode)

Dewhs committed:
feat(CheckBox): add animation when checked change

Just used this link https://developer.android.com/develop/ui/compose/animation/value-based

Dewhs committed:
fix(TextFieldLabel.typography):  bodyLarge -> labelLarge

Dewhs committed:
feat: custom navBar

RomanJos commented on android/app/src/main/java/com/example/makeat/ui/components/MCheckBox.kt:
```diff
@@ -30,12 +33,18 @@ fun MCheckbox(
     size: Dp = 24.dp
 ) {
     val colors = MaterialTheme.colorScheme
+    val animatedAlpha: Float by animateFloatAsState(
```
Bonne idée d'utiliser animateFloatAsState plutôt que animateColorAsState car t'as deux couleurs différentes derrière

RomanJos commented on android/app/src/main/java/com/example/makeat/ui/components/MNavBar.kt:
```diff
@@ -0,0 +1,69 @@
+import com.example.makeat.Stock
+import topBorder
+
+data class Item<T : Any>(val name: String, val route: T, val icon: ImageVector)
```
Ici tu pourrais virer les routes de ton appli et les mettres dans un dossier routes ou je sais pas mais laisser MNavBar dépourvu de ça, tu pourrais les utiliser seulement dans MainActivity.kt et instancier MNavBar dans le Scaffold avec un paramètres routes ou directement navItems quoi

Aussi, tu peut virer le Any soit par Composable si je comprends le truc ou soit en utilisant les [sealed classes](https://kotlinlang.org/docs/sealed-classes.html#constructors) et déclarer tes routes dedans car là tu peux passer une route qui est pas un composable et navController.navigate ou route::class.qualifiedName peut throw, c'est plus pour être dans la perfection car tu vas pas donner autre chose que tes routes mais voilà

RomanJos commented on android/app/src/main/java/com/example/makeat/utils/SideBorder.kt:
```diff
@@ -0,0 +1,58 @@
+}
+
+fun Modifier.rightBorder(
+    color: Color,
+    width: Float,
+) = this.drawWithContent {
+    drawContent()
+    drawLine(
+        color = color,
+        start = Offset(size.width, 0f),
+        end = Offset(size.width, size.height),
+        strokeWidth = width,
+    )
+}
+
+fun Modifier.bottomBorder(
+    color: Color,
+    height: Float,
+) = this.drawWithContent {
+    drawContent()
+    drawLine(
+        color = color,
+        start = Offset(0f, size.height),
+        end = Offset(size.width, size.height),
+        strokeWidth = height,
+    )
+}
+
+fun Modifier.leftBorder(
+    color: Color,
+    width: Float,
+) = this.drawWithContent {
+    drawContent()
+    drawLine(
+        color = color,
+        start = Offset(0f, 0f),
+        end = Offset(0f, size.height),
+        strokeWidth = width,
+    )
+}
```
```suggestion
// Google Gemini lol

import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * An enum to specify on which side a border should be drawn.
 */
enum class BorderSide { TOP, RIGHT, BOTTOM, LEFT }

/**
 * Draws a border on a single specified side of the Composable.
 *
 * @param side The side of the Composable to draw the border on.
 * @param width The width of the border, specified as a [Dp].
 * @param color The color of the border.
 */
fun Modifier.border(
    side: BorderSide,
    width: Dp = 1.dp,
    color: Color,
): Modifier = this.drawWithContent {
    drawContent()
    val strokeWidth = width.toPx()

    val (start, end) = when (side) {
        BorderSide.TOP ->    Offset(0f, 0f) to Offset(size.width, 0f)
        BorderSide.RIGHT ->  Offset(size.width, 0f) to Offset(size.width, size.height)
        BorderSide.BOTTOM -> Offset(0f, size.height) to Offset(size.width, size.height)
        BorderSide.LEFT ->   Offset(0f, 0f) to Offset(0f, size.height)
    }

    drawLine(
        color = color,
        start = start,
        end = end,
        strokeWidth = strokeWidth,
    )
}
```
Essaye ça mdr

RomanJos reviewed (approved):
Juste quelques amélioration mais sans plus

Dewhs committed:
fix(SideBorder): refactor code

Co-Authored-By: RomanJos <42180076+romanjos@users.noreply.github.com>

Dewhs commented on android/app/src/main/java/com/example/makeat/ui/LoginPage.kt:
```diff
@@ -78,7 +79,7 @@ fun LoginPage(modifier: Modifier = Modifier) {
             }
 
-            MButtonPrimary(text = "Enter", onClick = { TODO() /* Action login */ })
+            MButtonPrimary(text = "Enter", onClick = { onLoginSuccess() } )
```
Oui, ici, j'aurai le même comportement, mais uniquement parce que pour le moment, il n'y a aucune logique de login avec requêtes à l'API, etc. C'est pour ça que je le laisse comme ça pour le moment.

Dewhs commented on android/app/src/main/java/com/example/makeat/ui/components/MButtonPrimary.kt:
```diff
@@ -20,7 +19,7 @@ fun MButtonPrimary(
 ) {
     Button(
-        onClick = { onClick },
+        onClick = { onClick() },
```
Bien vu merci

Dewhs committed:
fix(MButtonPrimary.kt): { onClick() } -> onClick

use onClick's ref instead call a function which call onClick()

Co-authored-by: RomanJos <42180076+RomanJos@users.noreply.github.com>

Dewhs committed:
fix: refactor navigation

- Type Safe on NavItem data class thanks to sealed class Routes
- Dedicated navigation folder
- Move NavItem data class with instances from MNavBar.kt to NavItems.kt
- List<NavItem> is passed in param to MNavBar

RomanJos commented on android/app/src/main/java/com/example/makeat/MainActivity.kt:
```diff
@@ -34,53 +37,37 @@ class MainActivity : ComponentActivity() {
                     bottomBar = {
-                        if (currentDestination?.route != Login::class.qualifiedName) {
-                            MNavBar(navController)
+                        if (currentDestination?.route != Routes.Login::class.qualifiedName) {
```
Ben j'ai plus le truc pour faire des suggestions mais en gros fait 
if (!isLoginRoute) ou un autre nom pour clarifier car il y a un null check, ça prends une variable qui est ajouté par le décorateur etc tant ça peut être confusant (mais c'est un détail)

RomanJos commented on android/app/src/main/java/com/example/makeat/ui/components/MNavBar.kt:
```diff
@@ -59,15 +40,14 @@ fun MNavBar(navController: NavHostController, modifier: Modifier = Modifier) {
                 ),
```
Essaye de créer `NavigationBarItemDefaults.colors` dessus la for loop car c'est la même chose pour tous les items

Dewhs replied:
Je ne suis pas sûr que ça soit mieux de déclarer une variable juste au-dessus. Parce qu'en lisibilité, on n'y gagne pas vraiment, en optimisation, je ne suis pas certain non plus. Sachant qu'étant dans la boucle, on ne duplique pas de code non plus. Donc je suis plutôt mitigé.

RomanJos commented on android/app/src/main/java/com/example/makeat/ui/navigation/Routes.kt:
Faudrai tester de merger NavItem et ce fichier en mode
```kotlin

sealed class Routes {
    val label: String,
    val route: Routes,
    val icon: ImageVector,

    @Serializable
    data object Recipes : Routes() {
        "Recipes", this, ImageVector.vectorResource(R.drawable.chef_hat)
    }
```
(je sais pas si c'est possible mais tu vois l'idée)
Dans la navBar tu pourrai ensuite faire for (route in routes) et hop aura +1000

RomanJos reviewed (commented):
Des détails sauf pour les routes où j'aurais supprimé Drawable et mergé NavItem et Routes si possible tu verra

Par contre j'ai activé une nouvelle interface de code review sur github le truc il est un peu bugé lol je sais pas comment ça vas rendre lol

Dewhs commented on android/app/src/main/java/com/example/makeat/MainActivity.kt:
```diff
@@ -34,53 +37,37 @@ class MainActivity : ComponentActivity() {
                     bottomBar = {
-                        if (currentDestination?.route != Login::class.qualifiedName) {
-                            MNavBar(navController)
+                        if (currentDestination?.route != Routes.Login::class.qualifiedName) {
```
Je ne suis pas sûr de ce que tu veux dire. Tu veux juste dire sortir la condition pour lui donner un nom plus explicite ?
`val isLoginPage = currentDestination?.route == Routes.Login::class.qualifiedName`
et comme ça on as juste ton `if (!isLoginRoute)` ?

Dewhs committed:
style: move NavigationBarItemDefaults.colors out of for loop

Co-Authored-By: RomanJos <42180076+RomanJos@users.noreply.github.com>

Dewhs committed:
refactor: typed app icons

Co-Authored-By: RomanJos <42180076+RomanJos@users.noreply.github.com>

Dewhs committed:
feat: animated NavBar entry/exit

Co-Authored-By: RomanJos <42180076+RomanJos@users.noreply.github.com>
